File: iOS/Mobile/FeedFilterViewController.m
//
//  FeedFilterViewController.m
//  Mobile
//
//  Created by Jason Hocker on 8/17/12.
//  Copyright (c) 2012 Ellucian. All rights reserved.
//

#import "FeedFilterViewController.h"
#import "Feed.h"
#import "FeedCategory.h"
#import "UIViewController+GoogleAnalyticsTrackerSupport.h"

@interface FeedFilterViewController ()

@property (nonatomic,strong) NSSet *startingHiddenCategories;

@end

@implementation FeedFilterViewController
@synthesize navigationBar = _navigationBar;


-(void)viewDidLoad
{
    [super viewDidLoad];
    
    self.navigationController.navigationBar.translucent = NO;
    
    self.startingHiddenCategories = [self.hiddenCategories copy];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    [self initializeCategories];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return [self.categories count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *CellIdentifier = @"Feed Filter Cell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    NSString *category = [self.categories objectAtIndex:[indexPath row]];
    
    UILabel *textLabel = (UILabel *)[cell viewWithTag:1];
    textLabel.text = category;
    if([self.hiddenCategories containsObject:category]) {
        cell.accessoryType = UITableViewCellAccessoryNone;
    } else {
        cell.accessoryType = UITableViewCellAccessoryCheckmark;
    }
    return cell;
}

#pragma mark - Table view delegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    
    UITableViewCell *cell = [tableView cellForRowAtIndexPath:indexPath];
    UILabel *textLabel = (UILabel *)[cell viewWithTag:1];
    if(cell.accessoryType == UITableViewCellAccessoryCheckmark) {
        cell.accessoryType = UITableViewCellAccessoryNone;
        [self.hiddenCategories addObject:textLabel.text];
    } else {
        cell.accessoryType = UITableViewCellAccessoryCheckmark;
        [self.hiddenCategories removeObject:textLabel.text];
    }
    
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    if (_delegate) {
        [self updateCategories];
        [_delegate reloadData];
    }

}

- (void)updateCategories {
    
    NSError *error;
    if([self.hiddenCategories count] > 0) {
        self.feedModule.hiddenCategories = [[self.hiddenCategories allObjects] componentsJoinedByString:@","];
    } else {
        self.feedModule.hiddenCategories = nil;
    }
    
    if(![self.startingHiddenCategories isEqualToSet:self.hiddenCategories]) {
        [self sendEventToTracker1WithCategory:kAnalyticsCategoryUI_Action withAction:kAnalyticsActionList_Select withLabel:@"Filter changed" withValue:nil forModuleNamed:self.module.name];
    }
    [self.feedModule.managedObjectContext save:&error];
}

- (IBAction)dismiss:(id)sender {
    
    if(!_delegate)
    {
        [self updateCategories];
        [self dismissViewControllerAnimated:YES completion:nil];
    }
}

- (void)initializeCategories
{
    if( self.categories == nil )
    {
        NSError *error;
        NSFetchRequest *categoryRequest = [[NSFetchRequest alloc] init];
        [categoryRequest setEntity:[NSEntityDescription entityForName:@"FeedCategory" inManagedObjectContext:self.module.managedObjectContext ]];
        categoryRequest.predicate = [NSPredicate predicateWithFormat:@"moduleName = %@", self.module.name];
        NSMutableArray *categories = [[NSMutableArray alloc] init];
        NSArray *results = [self.module.managedObjectContext executeFetchRequest:categoryRequest error:&error ];
        for(FeedCategory *category in results) {
            [categories addObject:category.name];
        }
        self.categories = [[categories copy] sortedArrayUsingSelector:@selector(localizedCaseInsensitiveCompare:)];
    }
}

- (void)popoverControllerDidDismissPopover:(UIPopoverController *)popoverController
{
    if (_delegate)
    {
        [_delegate resetPopover];
    }
}

-(void)sizeForPopover
{
    //Make row selections persist.
    self.clearsSelectionOnViewWillAppear = NO;
    
    NSInteger rowsCount = [_categories count];
    NSInteger singleRowHeight = [self.tableView.delegate tableView:self.tableView
                                           heightForRowAtIndexPath:[NSIndexPath indexPathForRow:0 inSection:0]];
    NSInteger totalRowsHeight = rowsCount * singleRowHeight;
    
    //Calculate how wide the view should be by finding how
    //wide each string is expected to be
    CGFloat largestLabelWidth = 200;
    for (NSString *catName in _categories) {
        //Checks size of text using the default font for UITableViewCell's textLabel.
        CGSize labelSize = [catName sizeWithFont:[UIFont boldSystemFontOfSize:20.0f]];
        if (labelSize.width > largestLabelWidth) {
            largestLabelWidth = labelSize.width;
        }
    }
    
    //Add a little padding to the width
    CGFloat popoverWidth = largestLabelWidth + 100;
    
    //Set the property to tell the popover container how big this view will be.
    self.contentSizeForViewInPopover = CGSizeMake(popoverWidth, totalRowsHeight);
}


@end

File: iOS/Mobile/FeedFilterViewController.swift
//
//  FeedFilterViewController.swift
//  Mobile
//
//  Created by Jason Hocker on 7/31/15.
//  Copyright © 2015 Ellucian Company L.P. and its affiliates. All rights reserved.
//

import Foundation

protocol FeedFilterDelegate : class {
    func reloadData()
}

class FeedFilterViewController : UITableViewController {
    
    var startingHiddenCategories : NSMutableSet?
    var categories : NSArray?
    
    var feedModule : FeedModule?
    var hiddenCategories : NSMutableSet?
    var module : Module?
    weak var delegate : FeedFilterDelegate?
    
    override func viewDidLoad() {
        self.startingHiddenCategories = self.hiddenCategories
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        self.initializeCategories()
    }
    
    //MARK table view data source
    
    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 1
    }
    
    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.categories!.count
    }
    
    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCellWithIdentifier("Feed Filter Cell", forIndexPath: indexPath) as UITableViewCell
        
        let category = self.categories?.objectAtIndex(indexPath.row) as! String
        
        let textLabel = cell.viewWithTag(1) as! UILabel
        textLabel.text = category
        if (self.hiddenCategories!.containsObject(category) ) {
            cell.accessoryType = UITableViewCellAccessoryType.None
        } else {
            cell.accessoryType = .Checkmark
        }
        
        return cell
    }
    
    override func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        let cell = tableView.cellForRowAtIndexPath(indexPath)
        let textLabel = cell?.viewWithTag(1) as! UILabel
        
        if cell?.accessoryType == .Checkmark {
            cell?.accessoryType = .None
            self.hiddenCategories?.addObject(textLabel.text!)
        } else {
            cell?.accessoryType = .Checkmark
            //self.hiddenCategories = self.hiddenCategories!.filter() { $0 as? String != textLabel.text }
            self.hiddenCategories?.removeObject(textLabel.text!)
        }
        
        tableView.deselectRowAtIndexPath(indexPath, animated: true)
        
        self.updateCategories()
        self.delegate?.reloadData()
    }
    
    func updateCategories() {
        if hiddenCategories!.count > 0 {
            self.feedModule?.hiddenCategories = (hiddenCategories!.allObjects as! [String]).joinWithSeparator(",")
        } else {
            self.feedModule?.hiddenCategories = nil
        }
        
        if self.startingHiddenCategories!.isEqual(self.hiddenCategories) {
            sendEventToTracker1WithCategory(kAnalyticsCategoryUI_Action, withAction: kAnalyticsActionList_Select, withLabel: "Filter changed", withValue: nil, forModuleNamed: self.module?.name)
        }
        do {
            try feedModule?.managedObjectContext?.save()
        } catch {
            
        }
    }
    
    @IBAction func dismiss(sender: AnyObject) {
        updateCategories()
        dismissViewControllerAnimated(true, completion: nil)
    }
    func initializeCategories() {

        do {
            let request = NSFetchRequest(entityName: "FeedCategory")
            request.predicate = NSPredicate(format: "moduleName = %@", self.module!.name)
            let results = try self.module?.managedObjectContext?.executeFetchRequest(request) as! [FeedCategory]
            let categories = results.map {
                    return $0.name
            }
            self.categories = categories.sort { $0.localizedCaseInsensitiveCompare($1) == NSComparisonResult.OrderedDescending }
        } catch {
            
        }
    }
}